chore: Remove usage of deprecated `assertDictContainsSubset` unittest method.

lewisemm opened the pull request:
<!--

🌳🌳
🌳🌳🌳🌳         🌳 Note: Quince is in support. Fixes you make on master may still be needed on Quince.
    🌳🌳🌳🌳     If so, make another pull request against the open-release/quince.master branch,
🌳🌳🌳🌳         or ask in the #wg-build-test-release Slack channel if you have any questions or need help.
🌳🌳


Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

[`assertDictContainsSubset` has been deprecated since Python 3.2](https://docs.python.org/3.2/library/unittest.html#unittest.TestCase.assertDictContainsSubset) and [was removed in Python 3.12](https://docs.python.org/3/whatsnew/3.12.html#id3).

This PR removes all usage of `assertDictContainsSubset` and uses the set operator, `<=`,  on dictionary method `items()` for subset comparisons.

## Supporting information

https://github.com/openedx/public-engineering/issues/90

## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Include anything else that will help reviewers and consumers understand the change.
- Does this change depend on other changes elsewhere?
- Any special concerns or limitations? For example: deprecations, migrations, security, or accessibility.
- If your [database migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) can't be rolled back easily.

lewisemm committed:
chore: Fix deprecation warning from cms-1 tests

Fixes 1 deprecation warning from test runs on the cms-1 shard.

lewisemm committed:
chore: Fix deprecation warning from cms-2 tests

Fixes 1 deprecation warning from cms-2 test runs.

lewisemm committed:
chore: Fix deprecation logs from common-with-cms

Fix 3 deprecation warnings from test runs on common-with-cms shard.

lewisemm committed:
chore: Fix deprecation logs from common-with-lms

Fix 7 deprecation warnings from test runs on common-with-lms shard.

lewisemm committed:
chore: Fix deprecation logs from lms-1

Fix 4 deprecation warnings from test runs on lms-1 shard.

lewisemm committed:
chore: Fix deprecation logs from lms-3

Fix 10 deprecation warnings from test runs on lms-3 shard.

lewisemm committed:
chore: Fix deprecation logs from lms-4

Fix 6 deprecation warnings from test runs on lms-4 shard.

lewisemm committed:
chore: Fix deprecation logs from lms-5

Fix 1 deprecation warning from test runs on lms-5 shard.

lewisemm committed:
chore: Fix deprecation logs from lms-6

Fix 77 deprecation warnings from test runs on lms-6 shard.

lewisemm committed:
chore: Fix deprecation logs in openedx-1-with-cms

Fix 10 deprecation warnings from test runs on openedx-1-with-cms shard.

lewisemm committed:
chore: Fix deprecation logs in openedx-1-with-lms

Fix 2 deprecation warnings from test runs on openedx-1-with-lms shard.

lewisemm committed:
chore: Fix deprecation logs in openedx-2-with-cms

Fix 16 deprecation warnings from test runs on openedx-2-with-cms shard.

lewisemm committed:
chore: Fix deprecation logs in openedx-2-with-lms

Fix 30 deprecation warnings from test runs on openedx-2-with-lms shard.

lewisemm committed:
chore: Fix deprecation logs in xmodule-with-cms

Fix 3 deprecation warnings from test runs on xmodule-with-cms shard.

openedx-webhooks commented:
Thanks for the pull request, @lewisemm!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

mphilbrick211 commented:
Hi @lewisemm! Just flagging that there are a few failing tests.

lewisemm commented:
@mphilbrick211 Thanks for the nudge. Please allow me to convert this to draft as I figure out how to best fix those tests. Also, if you can send help my way that would be greatly appreciated.

mphilbrick211 commented:
Sounds good, @lewisemm! I'll look into getting you some help!

ormsbee commented on cms/djangoapps/contentstore/views/tests/test_block.py:
```diff
@@ -852,17 +852,17 @@ def test_duplicate_event(self):
+        }
+        self.assertTrue(
+            duplicated_event.items() <= event_receiver.call_args.kwargs.items()
         )
```
I believe you can do:

```python
assert duplicated_event.items() <= event_receiver.call_args.kwargs.items()
```

Or

```python
self.assertLessEqual(duplicated_event.items(), event_receiver.call_args.kwargs.items())
```

But trying use `assertTrue` here will cause pylint to flag it.

lewisemm replied:
This is perfect. I'll make those changes. Thank you.

lewisemm commented:
@mphilbrick211 Thanks. I got the help and I'll try to get this ready by end of week.

lewisemm committed:
chore: use unittest method assertLessEqual

Use assertLessEqual method to compare dictionary subset overlap.
This replaces the use of `assertTrue(dict1.items() <= dict2.items())`
format which raises pylint errors.

sarina commented:
@feanil would this PR be helpful for the python 3.12 upgrade?

feanil commented:
yes, we'll have to do it before we move edx-platform to Python 3.12

sarina commented:
If @lewisemm is still willing to contribute this fix, would you @feanil be able to work with them to get it merged?

feanil commented:
Yea, if they can update the PR to resolve conflicts, I'm happy to review and land the PR.

sarina commented:
Awesome, thank you!

@lewisemm - are you still interested in contributing this fix? If so it just needs to be updated to resolve conflicts.

lewisemm commented:
@sarina - Yes, I'm still interested in contributing this fix.

Hoping I can find some time to go over this within the next one week. Does that work with your release timelines?

sarina commented:
@lewisemm yes that will work well! Please ping @feanil when you're ready for a review.

lewisemm commented:
@feanil 

Update on this:

A lot has changed since I last submitted this PR. I have been trying to get the unit tests to run in order to get the latest deprecated messages but I'm facing two issues:

1. The unit tests are taking too long to run. For example, attempting to run `pytest common` [as documented here](https://docs.tutor.edly.io/tutorials/edx-platform.html#running-edx-platform-unit-tests) took at least two hours to get to 30% progress, after which I cancelled the test run.
2. A significant number of unit tests are failing, which means I cannot comprehensively address this issue.

I'd like to still have a go at this, but if time is of the essence then I have no problem with someone else taking over from here.

Also, if you might have some suggestions about how to work around the challenges I have listed above, that would be greatly helpful.

Thanks.

sarina commented:
@lewisemm thanks for the note. @feanil is out this week, but I'll make sure he gets back to you next week.

I think in the meantime if possible to rebase and resolve conflicts, we can see from the automated tests what's failing.

feanil commented:
@lewisemm Generally, I think it's okay to let the tests run in CI rather than trying to run them locally.  If you are running locally, I suggest only running the specific tests or test files that you're changing.   It would be great if you could update this branchand we can run the tests in GitHub which unually takes 15-25-mins

ChrisChV commented:
> I believe you can do:
> 
> ```python
> assert duplicated_event.items() <= event_receiver.call_args.kwargs.items()
> ```
> 
> Or
> 
> ```python
> self.assertLessEqual(duplicated_event.items(), event_receiver.call_args.kwargs.items())
> ```
> 
> But trying use `assertTrue` here will cause pylint to flag it.

Hi, what do you think about using `assertEqual(b, b | a)` ([ref](https://stackoverflow.com/questions/20050913/python-unittests-assertdictcontainssubset-recommended-alternative))? When the assertion fails, it shows a better log:

Log with `assertLessEqual`:

![Captura desde 2025-06-06 13-32-43](https://github.com/user-attachments/assets/42ef2ac7-9af7-4580-82bf-1f16db94a938)

Log with `assertEqual`

![Captura desde 2025-06-06 13-32-53](https://github.com/user-attachments/assets/863dc443-bf6c-4b55-a1fb-c22c9bc53b4f)

sarina commented:
@lewisemm - do you think you'll have the capacity to rebase this PR to get tests running?

lewisemm commented:
@sarina

Sorry, I will not be able to work on this. Please feel free to have someone else go at it.

Thanks.

sarina commented:
Thanks for letting us know @lewisemm and thank you for the initial contribution!

@feanil can you either take over or close this PR?

feanil commented:
I took a stab at rebasing this and unfortunately it got super complicated in the libraries code. I think it's going to make the most sense to just have someone create a new PR in the future rather than having me take over this one.
